apollo-router/src/plugins/authorization/scopes.rs

```diff
@@ -0,0 +1,672 @@
+            .cloned()
+            .collect::<HashSet<_>>();
+
+        self.request_scopes.is_superset(&type_scopes)
+    }
+}
+
+impl<'a> transform::Visitor for ScopeFilteringVisitor<'a> {
+    fn compiler(&self) -> &ApolloCompiler {
+        self.compiler
+    }
+
+    fn field(
+        &mut self,
+        parent_type: &str,
+        node: &hir::Field,
+    ) -> Result<Option<apollo_encoder::Field>, BoxError> {
+        let field_name = node.name();
+        let mut is_field_list = false;
+
+        let is_authorized = self
+            .compiler
+            .db
+            .types_definitions_by_name()
+            .get(parent_type)
+            .is_some_and(|def| {
+                if let Some(field) = def.field(&self.compiler.db, field_name) {
+                    if field.ty().is_list() {
+                        is_field_list = true;
+                    }
+                    self.is_field_authorized(field)
+                } else {
+                    false
+                }
+            });
+
+        self.current_path.push(PathElement::Key(field_name.into()));
+        if is_field_list {
+            self.current_path.push(PathElement::Flatten);
+        }
+
+        if !is_authorized {
+            self.unauthorized_paths.push(self.current_path.clone());
+        }
+
+        let res = if is_authorized {
+            transform::field(self, parent_type, node)
+        } else {
+            self.query_requires_scopes = true;
+            Ok(None)
+        };
+
+        if is_field_list {
+            self.current_path.pop();
+        }
+        self.current_path.pop();
+
+        res
+    }
+
+    fn fragment_definition(
+        &mut self,
+        node: &hir::FragmentDefinition,
+    ) -> Result<Option<apollo_encoder::FragmentDefinition>, BoxError> {
+        let fragment_is_authorized = self
+            .compiler
+            .db
+            .types_definitions_by_name()
+            .get(node.type_condition())
+            .is_some_and(|ty| self.is_type_authorized(ty));
+
+        if !fragment_is_authorized {
+            Ok(None)
+        } else {
+            transform::fragment_definition(self, node)
+        }
+    }
+
+    fn fragment_spread(
+        &mut self,
+        node: &hir::FragmentSpread,
+    ) -> Result<Option<apollo_encoder::FragmentSpread>, BoxError> {
+        let fragments = self.compiler.db.fragments(self.file_id);
+        let condition = fragments
+            .get(node.name())
+            .ok_or("MissingFragmentDefinition")?
+            .type_condition();
+        self.current_path
+            .push(PathElement::Fragment(condition.into()));
+
+        let fragment_is_authorized = self
+            .compiler
+            .db
+            .types_definitions_by_name()
+            .get(condition)
+            .is_some_and(|ty| self.is_type_authorized(ty));
+
+        // FIXME: if a field was removed inside a fragment definition, then we should add an unauthorized path
```

Comment:
this comment should be under fragment_definition

Reply:
the transform visitor handles operations before fragments. It should instead check fragments first, then operations. We would then be able to store the error paths per fragment, and at their application point, generate an error path starting from the operation